kratos/utilities/parallel_utilities.h

```diff
@@ -0,0 +1,202 @@
+inline void parallel_for(InputIt first, const int n, UnaryFunction &&f) {
+#pragma omp parallel for firstprivate(first)
+    for (int i = 0; i < n; ++i) {
+        f(first + i);
+    }
+}
+
+// parallel for executed by chunks. Here the function is called on every "block
+// of unknowns", so that the lambda captures are done just once
+// note that the function b must implement the loop from start to end
+// work allocation is static
+template <class TContainterType, class UnaryFunction>
+inline void parallel_for(TContainterType &data, UnaryFunction &&f) {
+    parallel_for(data.begin(), data.size(), f);
+}
+
+// the function f is executed in parallel. The parameter passed to f is the
+// number of the thread executing it.
+// one lambda capture is passed per every thread
+template <class UnaryFunction>
+inline void execute(UnaryFunction &&f) {
+    const int NumThreads = OpenMPUtils::GetNumThreads();
+
+#pragma omp parallel for firstprivate(NumThreads)
+    for (int i = 0; i < NumThreads; ++i) {
+        f(i);
+    }
+}
+
+// parallel for executed by chunks. Here the function is called on every "block
+// of unknowns", so that the lambda captures are done just once
+// note that the function b must implement the loop from start to end
+// work allocation is static
+template <class InputIt, class BinaryFunction>
+inline void block_parallel_for(InputIt first, const int n, BinaryFunction &&b) {
+    const int NumThreads = OpenMPUtils::GetNumThreads();
+    OpenMPUtils::PartitionVector Partitions;
+    OpenMPUtils::DivideInPartitions(n, NumThreads, Partitions);
+
+#pragma omp parallel for firstprivate(NumThreads)
+    for (int i = 0; i < NumThreads; ++i) {
+        b(first + Partitions[i], first + Partitions[i + 1]);
+    }
+}
+
+// parallel for executed by chunks. Here the function is called on every "block
+// of unknowns", so that the lambda captures are done just once
+// note that the function b must implement the loop from start to end
+// work allocation is dynamic
+template <class InputIt, class BinaryFunction>
+inline void block_parallel_for(
+    InputIt first, const int n, const int NChunks, BinaryFunction &&b) {
+    OpenMPUtils::PartitionVector Partitions;
+    OpenMPUtils::DivideInPartitions(n, NChunks, Partitions);
+
+#pragma omp parallel for firstprivate(first, NChunks) schedule(dynamic)
+    for (int i = 0; i < NChunks; ++i) {
+        b(first + Partitions[i], first + Partitions[i + 1]);
+    }
+}
+
+template <class TContainterType, class BinaryFunction>
+inline void block_parallel_for(TContainterType &data, BinaryFunction &&b) {
+    Kratos::Parallel::block_parallel_for(data.begin(), data.size(), b);
+}
+
+template <class TContainterType, class BinaryFunction>
+inline void block_parallel_for(
+    TContainterType &data, const int NChunks, BinaryFunction &&b) {
+    Kratos::Parallel::block_parallel_for(data.begin(), data.size(), NChunks, b);
+}
+
+inline void AtomicAdd(const double &value, double &target) {
```

Comment:
For readability some say it is better to do
```suggestion
inline void AtomicAdd(const double &value, double *target) {
```
since input/output relation is more visible in the calling code `AtomicAdd(a,&b)` than `AtomicAdd(a,b)`